### PR TITLE
Re-write end to end tests using cucumber and puppeteer

### PR DESCRIPTION
**Pull request by williamdes**
### Description

- Replace php selenium with browserstack to cucumber with puppeteer.
- I used puppeteer but you could use another selenium.

#### Benefits

- No programming language to write tests
- Quite fast
- Will be easy to ask contributors to add test files

### Websites
https://cucumber.io/docs/cucumber/
https://github.com/GoogleChrome/puppeteer

Ref #15154

### Progress

- [x] ChangePasswordTest.php
- [x] CreateRemoveUserTest.php
- [ ] ExportTest.php
- [x] LoginTest.php
- [ ] ServerSettingsTest.php
- [ ] TrackingTest.php
- [x] CreateDropDatabaseTest.php
- [ ] ImportTest.php
- [ ] NormalizationTest.php
- [ ] SqlQueryTest.php
- [x] TestBase.php
- [x] XssTest.php

#### Database
- [ ] EventsTest.php
- [ ] OperationsTest.php
- [ ] ProceduresTest.php
- [ ] QueryByExampleTest.php
- [ ] StructureTest.php
- [ ] TriggersTest.php

#### Table
- [ ] BrowseTest.php
- [ ] CreateTest.php
- [ ] InsertTest.php
- [ ] OperationsTest.php
- [ ] StructureTest.php

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,23 +3,15 @@ dist: xenial
 language: php
 
 stages:
+  - name: "Other tests"
   - name: "Lint and analyse code"
   - name: "PHP Unit tests"
   - name: "Documentation"
-  - name: "Other tests"
```
Comment (williamdes) [inline .travis.yml]: Temporary change to not waste resources when developing and testing scenarios on TravisCI

**Comment by williamdes** [comment]
@mauriciofauth @ibennetch can you please review the PR
It is not finished but I want your approval or some advice before converting more scenarios.
Thanks a lot :)

**Comment by devenbansod** [comment]
@williamdes thanks for this!

This is a great initiative. This would definitely help us to source more end-to-end tests from the new PRs. And if these tests turn out to be more stable that current ones, that'd be a cherry-on-top 👍 🥇 💯 

Also, I haven't tested these changes locally but does this change require any change in the local Selenium setup?

**Comment by williamdes** [comment]
@devenbansod what I really like is that there is not setup
Only 3 ENVs (the phpmyadmin http host, the user, the password)
Puppeteer will download a chromium when you do npm install

Thanks for the review! :)

**Review by MauricioFauth — APPROVED** [review]
Nice work!
Could you remove Selenium in a different PR?
Is it possible to add a command to run these tests in a non-headless browser like Selenium?

File: test/end-to-end/features/support/world.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    scope.password = process.env.TESTSUITE_PASSWORD || '';
+    scope.windowWidth = 1024;
+    scope.windowHeight = 800;
+};
+
+setWorldConstructor(World);
+
+const browsePage = async (path) => {
+    var headless = false;
+    if (process.env.TESTSUITE_HEADLESS) {
```
Comment (williamdes) [inline test/end-to-end/features/support/world.js]: @mauriciofauth Here you can have a GUI or use headless mode

I will separate selenium tests removal in another PR

Comment (mauriciofauth) [reply test/end-to-end/features/support/world.js]: Thanks! :+1:

**Comment by williamdes** [comment]
To add more fun and Tu be sure I cover most of the code I will add a way to measure code coverage for php
https://tarunlalwani.com/post/php-code-coverage-web-selenium/

And for the js part 
I will try the tool https://github.com/istanbuljs/puppeteer-to-istanbul

**Comment by williamdes** [comment]
Nope

**Comment by williamdes** [comment]
```bash
TESTSUITE_URL="http://pma.selenium.local/" TESTSUITE_PASSWORD="public" yarn end-to-end
```

```php
if (($_SERVER['HTTP_HOST'] ?? '') === 'pma.selenium.local') {
    $cfg['DBG']['sql'] = false;
    $cfg['DBG']['sqllog'] = false;
    $cfg['DBG']['demo'] = false;
    $cfg['DBG']['simple2fa'] = false;
    $cfg['FilterLanguages'] = '^(en)$';
    $cfg['Lang'] = 'en';
    $i = 0;
    $cfg['Servers'] = [];
    $i++;
    $cfg['Servers'][$i]['verbose'] = 'Local';
    $cfg['Servers'][$i]['auth_type'] = 'cookie';
    $cfg['Servers'][$i]['host'] = 'mysql80.williamdes.local';
    $cfg['Servers'][$i]['port'] = 3306;
    $cfg['Servers'][$i]['AllowNoPassword'] = true;
    $cfg['Servers'][$i]['compress'] = false;
}
```

**Comment by williamdes** [comment]
```bash
yarn global add selenium-standalone
~/.yarn/bin/selenium-standalone start --  -port 5556
```

Actual suite
```
TESTSUITE_SELENIUM_HOST="127.0.0.1" TESTSUITE_SELENIUM_PORT="5556" TESTSUITE_SERVER="172.23.0.17" TESTSUITE_USER="root" TESTSUITE_URL="http://pma.selenium.local/" TESTSUITE_PASSWORD="public" ./vendor/bin/phpunit --group=selenium
```

**Comment by williamdes** [comment]
Closing for now, cucumber will maybe be an option but for now #16134 cleaned part of the trouble.
Thank you for all the comments made on this PR !

**Pull request closed without merging**